### PR TITLE
Use an OFD lock to prevent more than one fwupdtool process

### PR DESCRIPTION
**Pull request by hughsie**
Fixes https://github.com/fwupd/fwupd/issues/3019

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

File: src/fu-tool.c [hunk 1]
```diff
@@ -2781,6 +2784,48 @@ fu_util_switch_branch (FuUtilPrivate *priv, gchar **values, GError **error)
...
+	struct flock lockp = {
+		.l_type = F_WRLCK,
+		.l_whence = SEEK_SET,
+	};
+	g_autofree gchar *localstatedir = NULL;
+	g_autofree gchar *lockfn = NULL;
+
+	/* open file */
+	localstatedir = fu_common_get_path (FU_PATH_KIND_LOCALSTATEDIR);
+	lockfn = g_build_filename (localstatedir, "run", "fwuptool", NULL);
```
Comment (superm1) [inline src/fu-tool.c:2800]: `s/fwuptool/fwupdtool/`

File: src/fu-tool.c [hunk 2]
```diff
@@ -2781,6 +2784,48 @@ fu_util_switch_branch (FuUtilPrivate *priv, gchar **values, GError **error)
...
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "failed to open %s",
+			     lockfn);
+		return FALSE;
+	}
+
+	/* write lock */
+	if (fcntl (priv->lock_fd, F_OFD_SETLK, &lockp) < 0) {
```
Comment (superm1) [inline src/fu-tool.c:2814]: I suppose it's not actually required to set `F_UNLCK` on close because the descriptor releases, right?

Comment (hughsie) [reply src/fu-tool.c:2814]: Yup, that's why I chose ofd rather than a file lock.

Comment (superm1) [reply src/fu-tool.c:2814]: Sorry, I should fully complete my thoughts - if the fwupdtool crashes I want to make sure that we don't end up with dangling descriptors, which I don't know if that's a thing.

**Review by superm1 — APPROVED** [review]

File: src/fu-tool.c (hunk 1 above)
Comment (hughsie) [reply src/fu-tool.c:2800]: Thanks, fixed. I think the tool should probably have been called fwuptool and fwupmgr but we can't go back in time now :)

**Comment by djcampello** [comment]
Does it make sense to allow an option to busy wait for the lock to be released? I'm not sure is a great idea as there could be a stale lock but I'm trying to figure out the best way to serialize concurrent executions...

**Comment by hughsie** [comment]
> Does it make sense to allow an option to busy wait for the lock to be released

From a "_fwupdtool is a development tool_" point of view I think exiting straight away makes sense. From a "_launching fwupdtool from an upstart script_" point of view maybe it does make sense to just wait for the lock to clear. If you need the latter I'd be fine reviewing it, but I've got tons of stuff on my plate for the next few days.

I chose OFD locks as it's impossible for them to be stale as the kernel cleans up after whatever userspace does.

**Comment by superm1** [comment]
I almost wonder if the upstart fwupdtool scenario should have it's _own_ lock and that behavior.

**Pull request merged**